code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServerImpl.java

```diff
@@ -0,0 +1,296 @@
+        String dockerComposeDataStr = yaml.dump(dockerComposeData);
+        String rancherComposeDataStr = yaml.dump(rancherComposeData);
+        return new SimpleEntry<String, String>(dockerComposeDataStr, rancherComposeDataStr);
+    }
+
+    private Map<String, Object> createComposeData(List<? extends Service> services, boolean forDockerCompose) {
+        Map<String, Object> data = new HashMap<String, Object>();
+        for (Service service : services) {
+            Map<String, Object> rancherServiceData = getServiceDataAsMap(service);
+            Map<String, Object> composeServiceData = new HashMap<>();
+            for (String rancherService : rancherServiceData.keySet()) {
+                ServiceDiscoveryConfigItem item = ServiceDiscoveryConfigItem.getServiceConfigItemByRancherName(rancherService);
+                if (item != null && item.isDockerComposeProperty() == forDockerCompose) {
+                    Object value = rancherServiceData.get(rancherService);
+                    boolean export = false;
+                    if (value instanceof List) {
+                        if (!((List<?>) value).isEmpty()) {
+                            export = true;
+                        }
+                    } else if (value instanceof Map) {
+                        if (!((Map<?, ?>) value).isEmpty()) {
+                            export = true;
+                        }
+                    } else if (value != null) {
+                        export = true;
+                    }
+                    if (export) {
+                        composeServiceData.put(item.getComposeName(), value);
+                    }
+                }
+            }
+            
+            if (forDockerCompose) {
+                populateLinksForService(service, services, composeServiceData);
+                populateVolumesForService(service, services, composeServiceData);
+            }
+            
+            data.put(service.getName(), composeServiceData);
+        }
+        return data;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void populateVolumesForService(Service service, List<? extends Service> servicesToExport,
+            Map<String, Object> composeServiceData) {
+        List<String> namesCombined = new ArrayList<>();
+        List<String> servicesNames = new ArrayList<>();
+        List<Service> translateToInstances = new ArrayList<>();
+        List<? extends Integer> consumedServiceIds = (List<? extends Integer>) getServiceDataAsMap(service).get(
+                ServiceDiscoveryConfigItem.VOLUMESFROMSERVICE.getRancherName());
+
+        if (consumedServiceIds != null) {
+            for (Integer consumedServiceId : consumedServiceIds) {
+                Service consumedService = objectManager.findOne(Service.class, SERVICE.ID,
+                        consumedServiceId);
+                boolean found = false;
+                for (Service serviceToExport : servicesToExport) {
```

Comment:
Could be worth it to create a lookup map to use prior to iterating through consumedServiceIds.  Not sure how large these lists can get.